Add hook to initialize Julia on-the-fly during thread adoption

gbaraldi opened the pull request:
~The big changes are moving initialization to when someone calls into julia, making it all automatic. (It still supports manually initializing the runtime). I'm not done with this because i still need to figure out how to embed runtime arguments. Currently it's piped as a string but it's not clear yet if that's the best way forward.~

~We maybe want to make the changes to adopt thread handling only exist behind a flag. Though that potentially doesn't add much because essentially this adds a very predictable null check to ccallable~

**Edit:** This PR has been partially split. It no longer makes any changes to PTLS codegen, and instead just adds `jl_pgcstack_default_func` to allow an uninitialized runtime to make it into `jl_autoinit_and_adopt_thread`

The init changes themselves are left to #57498

topolarity commented on src/jlapi.c:
```diff
@@ -1073,6 +1088,10 @@ JL_DLLEXPORT int jl_repl_entrypoint(int argc, char *argv[])
     }
 
     julia_init(jl_options.image_file_specified ? JL_IMAGE_CWD : JL_IMAGE_JULIA_HOME);
+    jl_atomic_store_release(&jl_runtime_is_initialized, 1);
```
Should this be set by `julia_init` instead?

It seems like embedding applications might call that directly

topolarity commented on src/llvm-ptls.cpp:
```diff
@@ -169,17 +182,29 @@ void LowerPTLS::fix_pgcstack_use(CallInst *pgcstack, Function *pgcstack_getter,
+                    XSTR(jl_find_dynamic_library_by_addr), M);
+            }
+            auto this_func = builder.GetInsertBlock()->getParent();
+            handle = builder.CreateCall(dladdr, {ConstantExpr::getBitCast(this_func, builder.getPtrTy())});
```
Do we ever use this in pkgimages?

It seem like it will get the wrong handle in that case.

gbaraldi replied:
No, unless you are loading a package image as a sysimage in which case glhf

topolarity commented on src/llvm-ptls.cpp:
```diff
@@ -143,7 +146,17 @@ void LowerPTLS::fix_pgcstack_use(CallInst *pgcstack, Function *pgcstack_getter,
+        // if (offset != 0)
+        //     pgcstack = tp + offset; // fast
+        // else
+        // pgcstack_getter = load pgcstack_func_slot
```
This pseudocode doesn't look properly indented

topolarity commented:
Looks like this ironically brokes the trimming test.

Any idea what's up @gbaraldi ?

topolarity commented on src/jlapi.c:
```diff
@@ -1126,6 +1145,29 @@ JL_DLLEXPORT int jl_repl_entrypoint(int argc, char *argv[])
+    }
+    //TODO: Implement option parsing
+    assert(sysimg_handle);
+    // This code path assumes that you are either an executable or dylib with the sysimage linked in
```
Is that a fair assumption?

topolarity commented on src/init.c:
```diff
@@ -869,9 +870,10 @@ static NOINLINE void _finish_julia_init(JL_IMAGE_SEARCH rel, jl_ptls_t ptls, jl_
     if (rel == JL_IMAGE_IN_MEMORY) {
-        jl_set_sysimg_so(jl_exe_handle);
-        jl_options.image_file = jl_options.julia_bin;
+        jl_set_sysimg_so(jl_base_image_handle);
+        jl_options.image_file = jl_options.julia_bin; // XXX: Use dladdr on base_image_handle?
```
It is possible to pass this in explicitly instead of relying on it to be initialized externally?

gbaraldi committed:
Implement code for initialization of julia from adopt thread

topolarity reviewed (approved):
Thanks for splitting this for easy review!

I'm not a huge fan of the breaking change to `jl_adopt_thread` (a public, albeit dangerous-to-use function) - I think it's confusing that it now asks you for a parameter it almost never uses.

Other than that, changes look good to me 👍 thanks @gbaraldi

vtjnash reviewed (commented):
I am not very happy about injecting all of this new code into every cfunction we have. I think you need to actually initialize `pgcstack_func_slot` with something

gbaraldi commented:
Yeah, I was thinking of some way of making this optional on sysimages. I guess we can add an attr to the sysimage module?

topolarity commented:
> I am not very happy about injecting all of this new code into every cfunction we have.

Is your concern code size, or performance overhead for the extra branch?

vtjnash commented:
Size, affecting speed and maintainability if we add other checks and atomics eventually, for what is always the same slow path code

topolarity committed:
Re-work `jl_adopt_thread` runtime auto-initialization

This should be a dramatically simpler implementation, by just setting up
the `jl_pgcstack_func_slot` to have a usable function pointer by default
and then performing library introspection based on
`__builtin_return_address` in the runtime.

This avoids needing to modify codegen for `julia.get_pgcstack` at all.

topolarity commented:
I took the liberty of pushing a much simpler strategy of just initializing `jl_pgcstack_func_slot` and using `__builtin_return_address` for figuring out the library / entrypoint that triggered auto-init.

This no longer requires changes to PTLS codegen
